### ISSUE TITLE
Require valid Content-Type and Accept headers

### ISSUE THREAD

**Issue by BavoNootaert** [erasmus-without-paper/ewp-specs-architecture#20]
We're getting errors in our application when processing the 'fake api's'  that just specify files from github. The reason is that they have Content-Type 'text/plain' instead of 'application/xml'. I haven't found anything in the EWP specs about the content-type. Should we be able to process any content that might be xml? Or should application/xml be mandatory?

**Comment by wrygiel** [comment]
Interesting question.

The specifications (currently) don't mention that, but I think it's a common knowledge that `Content-Type` "SHOULD" be set correctly. But should we make it a "MUST"? Perhaps.

BTW, `text/xml` also seems to be a valid `Content-Type` for most of our responses (so we would have to allow at least two).

**Comment by wrygiel** [comment]
> We're getting errors (...) The reason is that they have Content-Type 'text/plain' instead of 'application/xml'.

`Content-Type` is usually ignored by the client code (because the application already knows what to expect). Are you sure that this is the reason? Perhaps you are making your request with the POST method, and GitHub responds with HTTP 400 (because it doesn't allow POST)?

I will remove "fake APIs" from dev-registry anyway, but please investigate a little more.

**Comment by BavoNootaert** [comment]
We make a GET request, and GitHub returns an HTTP 200 response. The library we're using explicitly checks the content type.

text/plain may be exceptional. But the problem goes a bit deeper. E.g. other REST API's sometimes use API-specific content-types (like application/vnd...+xml). The idea being that the same resource can be queried with different XML-formats. MAY an EWP implementation use such a content-type, and SHOULD all clients be able to process those responses, even if for now we don't actually use different representations for the same resource?

**Comment by wrygiel** [comment]
Okay, I understand.

Yes, many APIs vary their response type on the request's `Accept` header. Currently EWP doesn't use this feature, and - probably - won't use it, for the reasons of backward-compatibility. But you're right - if some libraries expect specific content types, then I think we should make it a MUST.

**Comment by wrygiel** [comment]
This topic is also related to the newly introduced (in Porto) topic of "double encryption". If the response is encrypted, it's content-type might be replaced with `application/octet-stream`.

**Comment by wrygiel** [comment]
I think we should sit on this for a little longer. It's hard to guess at this point how encryption will affect this topic.

**Comment by wrygiel** [comment]
@mpuzar, @richared, this is the topic @BavoNootaert brought up during the meeting today.

**Comment by mpuzar** [comment]
I have just checked on one of our APIs, and we do serve the correct content type ("application/xml"). If I understood correctly, it was mentioned that Oslo is serving it wrong...? If that is indeed the case then I need more information on where/when this is happening. It would be good to have it right anyway, because it will take time to implement the encryption functionality.

As for application/octet-stream, this would depend on how we choose to transport the data. If it is base64 encoded then that would not be necessary.

**Comment by BavoNootaert** [comment]
@mpuzar, you are indeed serving the correct content type, for the Echo Institutions and Organizational Units APIs. I hadn't rechecked before the meeting. Sorry about that.

**Comment by wrygiel** [comment]
> This topic is also related to the newly introduced (in Porto) topic of "double encryption". If the response is encrypted, it's content-type might be replaced with `application/octet-stream`.

Since `v0.3.0`, both request and response encryption specs (the `ewp-rsa-aes128gcm` ones) require the sender to leave `Content-Type` "as is".

**Comment by wrygiel** [comment]
BTW - if we make it a MUST for the server to supply a valid `Content-Type`, then I believe we also should make it a MUST for the client to supply a proper `Accept` header.

**Comment by wrygiel** [comment]
Note, that this change is not strictly backward-compatible (however, one might argue about this). Do all partners agree with this approach?

@erasmus-without-paper/all-members

**Comment by mikesteez** [comment]
It is common for a REST-API to respond with HTTP 406 Not Acceptable if the client requests a media type which is not supported. 

http://restlet.com/company/blog/2015/12/10/understanding-http-content-negotiation/

**Comment by wrygiel** [comment]
> It is common for a REST-API to respond with HTTP 406 Not Acceptable if the client requests a media type which is not supported.

Yes. This would work similar as `Accept-Encoding` and `Content-Encoding` work in [response encryption](https://github.com/erasmus-without-paper/ewp-specs-sec-resencr-rsa-aes128gcm).

**Comment by wrygiel** [comment]
Please review this PR: https://github.com/erasmus-without-paper/ewp-specs-architecture/pull/28

Will this be enough? I'd to merge it tomorrow.

**Comment by BavoNootaert** [comment]
Perhaps instead of 'most often, this will be `text/xml` or  `application/xml`', it should be 'unless otherwise noted in the API, this will be `text/xml` or `application/xml`'. Or something like that.

**Comment by wrygiel** [comment]
> Perhaps instead of 'most often, this will be `text/xml` or  `application/xml`', it should be 'unless otherwise noted in the API, this will be `text/xml` or `application/xml`'. Or something like that.

I think that the content-type is - in most cases - self-apparent (the sender should know it by himself, without our help). We just need to make sure that he knows that we care.

Also, if we did what you suggest, then - if we take this advice literally - we would be required to annotate every non-XML resource saying that it's not XML (for example, logo image URL in the Institutions API).

**Comment by BavoNootaert** [comment]
I don't think that how the logo image URL (or the fact sheet URL or the website URL) in the Institutions API is handled is part of the EWP API. Only the URL's that appear in the manifest, are.

**Comment by wrygiel** [comment]
Perhaps. But this would require us to specify what we still consider as "API responses" and what not.

You didn't reply on my first argument (the one in which I claim that a valid content-type should be self-apparent, and we don't need to specify it explicitly). Do you think it's okay of us to assume that?

**Comment by BavoNootaert** [comment]
With lots of partners, there will be several opinions about the most obvious content-type. As indicated earlier, someone may choose 'application/vnd.myewpimplementation.institutions+xml', and your proposal would allow that. But some partners may not support that, leading to discussions (of which the end-users will be the victim) that could have been avoided by just being clear.

'Unless otherwise noted, the content-type for xml-based API's MUST be text/xml or application/xml'?

**Comment by wrygiel** [comment]
We talked with Bavo in Warsaw. The safest thing to do here would to be specify this explicitly in every API specification. E.g. say that the response needs to have this and that Content-Type.

(Alternative would be explicitly specify that the response can have any Content-Type, but we agreed that this is a slightly worse option.)
